core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java

```diff
@@ -645,12 +645,67 @@ public void testStressAcquireRelease()
 
         try (NodeAllocator nodeAllocator = nodeAllocatorService.getNodeAllocator(SESSION)) {
             for (int i = 0; i < 10_000_000; ++i) {
-                NodeAllocator.NodeLease lease = nodeAllocator.acquire(REQ_NONE, DataSize.of(32, GIGABYTE));
+                NodeAllocator.NodeLease lease = nodeAllocator.acquire(REQ_NONE, DataSize.of(32, GIGABYTE), false);
                 lease.release();
             }
         }
     }
 
+    @Test(timeOut = TEST_TIMEOUT)
+    public void testAllocateSpeculative()
+    {
+        InMemoryNodeManager nodeManager = new InMemoryNodeManager(NODE_1, NODE_2);
+        setupNodeAllocatorService(nodeManager);
+
+        try (NodeAllocator nodeAllocator = nodeAllocatorService.getNodeAllocator(SESSION)) {
+            // allocate two speculative tasks
+            NodeAllocator.NodeLease acquireSpeculative1 = nodeAllocator.acquire(REQ_NONE, DataSize.of(64, GIGABYTE), true);
+            assertAcquired(acquireSpeculative1, NODE_1);
+            NodeAllocator.NodeLease acquireSpeculative2 = nodeAllocator.acquire(REQ_NONE, DataSize.of(32, GIGABYTE), true);
+            assertAcquired(acquireSpeculative2, NODE_2);
+
+            // non-speculative tasks should still get node
+            NodeAllocator.NodeLease acquireNonSpeculative1 = nodeAllocator.acquire(REQ_NONE, DataSize.of(64, GIGABYTE), false);
+            assertAcquired(acquireNonSpeculative1, NODE_2);
+            NodeAllocator.NodeLease acquireNonSpeculative2 = nodeAllocator.acquire(REQ_NONE, DataSize.of(32, GIGABYTE), false);
+            assertAcquired(acquireNonSpeculative2, NODE_1);
```

Comment:
For my understanding, two nodes have a total of 128GB, but now we have 192GB scheduled (96GB speculative, 96GB non-speculative), right? How does this work out later, scheduler will preempt speculative tasks in favor of non-speculative tasks?

Reply:
> How does this work out later, scheduler will preempt speculative tasks in favor of non-speculative tasks

They will work side by side until we run out of memory on node. In such case speculative task will be killed by low memory killer. If they will fit in available memory (task does not alway use all reserved memory) both should complete succesfully.